crates/core/tedge_mapper/src/c8y/converter.rs

```diff
@@ -515,16 +547,91 @@ fn create_get_pending_operations_message() -> Result<Message, ConversionError> {
 fn create_supported_operations_fragments_message(
     cfg_dir: &Path,
 ) -> Result<Message, ConversionError> {
-    let ops_dir = format!("{}/{SUPPORTED_OPERATIONS_DIRECTORY}", cfg_dir.display());
-    let ops = Operations::try_new(ops_dir, C8Y_CLOUD)?;
+    let dir_split: Vec<&str> = cfg_dir
+        .to_str()
+        .ok_or_else(|| ConversionError::DirPathComponentError {
+            dir: cfg_dir.to_path_buf(),
+        })?
+        .split('/')
+        .collect();
+
+    if dir_split.len() == 6 {
+        let stopic = format!("{SMARTREST_PUBLISH_TOPIC}/{}", dir_split[5]);
+
+        create_ops_message(Operations::try_new(cfg_dir)?, Topic::new_unchecked(&stopic))
+    } else {
+        let ops_dir = format!(
+            "{}/{SUPPORTED_OPERATIONS_DIRECTORY}/{C8Y_CLOUD}",
+            cfg_dir.display()
+        );
+        create_ops_message(
+            Operations::try_new(ops_dir)?,
+            Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC),
+        )
+    }
+}
+
+fn create_ops_message(ops: Operations, topic: Topic) -> Result<Message, ConversionError> {
     let ops = ops.get_operations_list();
     let ops = ops.iter().map(|op| op as &str).collect::<Vec<&str>>();
-
     let ops_msg = SmartRestSetSupportedOperations::new(&ops);
-    let topic = Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC);
+
     Ok(Message::new(&topic, ops_msg.to_smartrest()?))
 }
 
+fn create_child_supported_operations_fragments_message(
+    children: &mut HashSet<String>,
+    cfg_dir: &Path,
+) -> Result<Vec<Message>, ConversionError> {
+    let mut mqtt_messages = Vec::new();
+    let ops_dir = format!("{}/{SUPPORTED_OPERATIONS_DIRECTORY}", cfg_dir.display());
+    let path: PathBuf = ops_dir.into();
+    let child_entries = fs::read_dir(&path.join(C8Y_CLOUD))
+        .map_err(|_| ConversionError::ReadDirError {
+            dir: PathBuf::from(&path),
+        })?
+        .map(|entry| entry.map(|e| e.path()))
+        .collect::<Result<Vec<PathBuf>, _>>()?
+        .into_iter()
+        .filter(|path| path.is_dir())
+        .collect::<Vec<PathBuf>>();
+
+    for cdir in child_entries {
+        create_mqtt_message_for_child_device(cdir, children, &mut mqtt_messages)?;
+    }
+    Ok(mqtt_messages)
+}
+
+// Check if the child is already created or not
+// If not create the child creation message and then the operations message.
+fn create_mqtt_message_for_child_device(
+    cdir: PathBuf,
+    children: &mut HashSet<String>,
+    mqtt_messages: &mut Vec<Message>,
+) -> Result<(), ConversionError> {
+    let ops = Operations::try_new(cdir.clone())?;
+    let ops = ops.get_operations_list();
+    let ops = ops.iter().map(|op| op as &str).collect::<Vec<&str>>();
+    let ops_msg = SmartRestSetSupportedOperations::new(&ops);
+    if let Some(id) = cdir.file_name() {
+        if id.to_str().is_some() {
```

Comment:
This need to be removed. `if x.is_some()` is a misused of the `Option` type.

Reply:
fixed